File: platform/src/lib/algorithm.ts
import type { ApproachLive, Counts, LightColor, Mode, TimingProfile } from "./types";

export const DEFAULT_TIMING: TimingProfile = {
  minGreenS: 8,
  maxGreenS: 60,
  yellowS: 3,
  allRedS: 1,
  minPedS: 12,
  extensionS: 3,
};

export function emptyCounts(): Counts {
  return {
    m100: 0,
    m200: 0,
    m300: 0,
    waitS: 0,
    motos: 0,
    cars: 0,
    buses: 0,
    trucks: 0,
    peds: 0,
    emergency: false,
  };
}

export function approachScore(c: Counts, mode: Mode): number {
  let score =
    c.m100 * 1.0 +
    c.m200 * 1.6 +
    c.m300 * 2.4 +
    c.waitS * 0.08 +
    c.motos * 0.45 +
    c.buses * 2.2 +
    c.trucks * 2.8 +
    c.peds * 1.8;
  if (mode === "school") score += c.peds * 2.5;
  if (mode === "night" && c.m100 + c.m200 + c.m300 === 0) score *= 0.2;
  if (c.emergency) score += 10_000;
  return Number(score.toFixed(1));
}

function axisOf(heading: number): "ns" | "ew" {
  const d = ((heading % 360) + 360) % 360;
  return d === 90 || d === 270 ? "ew" : "ns";
}

function assertNoConflict(approaches: ApproachLive[]): ApproachLive[] {
  const greens = approaches.filter((a) => a.color === "green");
  const axes = new Set(greens.map((g) => axisOf(g.headingDeg)));
  if (axes.size > 1) {
    return approaches.map((a) => ({ ...a, color: "flashing_amber" as LightColor }));
  }
  return approaches;
}

function setAxisGreen(approaches: ApproachLive[], axis: "ns" | "ew"): ApproachLive[] {
  return approaches.map((a) => ({
    ...a,
    color: (axisOf(a.headingDeg) === axis ? "green" : "red") as LightColor,
    greenElapsedS: axisOf(a.headingDeg) === axis ? a.greenElapsedS : 0,
  }));
}

export function tickPhase(
  approaches: ApproachLive[],
  mode: Mode,
  timing: TimingProfile = DEFAULT_TIMING,
): ApproachLive[] {
  const scored = approaches.map((a) => ({
    ...a,
    score: approachScore(a.counts, mode),
  }));

  if (mode === "failsafe") {
    return scored.map((a) => ({ ...a, color: "flashing_amber" as LightColor }));
  }

  const emergency = scored.find((a) => a.counts.emergency);
  if (emergency) {
    return assertNoConflict(setAxisGreen(scored, axisOf(emergency.headingDeg)));
  }

  const greens = scored.filter((a) => a.color === "green");
  const currentAxis: "ns" | "ew" =
    greens.length > 0 ? axisOf(greens[0].headingDeg) : "ns";

  if (greens.length === 0) {
    const winner = [...scored].sort((a, b) => b.score - a.score)[0];
    return assertNoConflict(
      setAxisGreen(
        scored.map((a) => ({ ...a, greenElapsedS: 0 })),
        axisOf(winner.headingDeg),
      ),
    );
  }

  const group = scored.filter((a) => axisOf(a.headingDeg) === currentAxis);
  const otherGroup = scored.filter((a) => axisOf(a.headingDeg) !== currentAxis);
  const groupScore = group.reduce((s, a) => s + a.score, 0);
  const otherScore = otherGroup.reduce((s, a) => s + a.score, 0);
  const elapsed = Math.max(...group.map((a) => a.greenElapsedS)) + 1;
  const empty = group.every(
    (a) => a.counts.m100 + a.counts.m200 + a.counts.m300 === 0,
  );
  const servedMin = elapsed >= timing.minGreenS;
  const hitMax = elapsed >= timing.maxGreenS;
  const pedHold = group.some((a) => a.pedWaiting) && elapsed < timing.minPedS;
  const incoming = group.some((a) => a.counts.m200 > 0 || a.counts.m300 > 0);
  const shouldCut = servedMin && empty && otherScore > 2;

  if (pedHold) {
    return assertNoConflict(
      scored.map((a) =>
        axisOf(a.headingDeg) === currentAxis ? { ...a, greenElapsedS: elapsed } : a,
      ),
    );
  }

  if (shouldCut || (hitMax && otherScore >= groupScore)) {
    const nextAxis = currentAxis === "ns" ? "ew" : "ns";
    return assertNoConflict(
      setAxisGreen(
        scored.map((a) => ({ ...a, greenElapsedS: 0 })),
        nextAxis,
      ),
    );
  }

  return assertNoConflict(
    scored.map((a) =>
      axisOf(a.headingDeg) === currentAxis
        ? {
            ...a,
            greenElapsedS: incoming && !empty ? elapsed + timing.extensionS : elapsed,
          }
        : a,
    ),
  );
}

export function modeLabel(mode: Mode): string {
  const map: Record<Mode, string> = {
    normal: "Adaptativo",
    school: "Colegio",
    market: "Mercado",
    night: "Noche segura",
    eco: "Eco solar",
    emergency: "Emergencia",
    failsafe: "Ámbar seguro",
  };
  return map[mode];
}

File: firmware/edge_controller/algorithm.py
"""SmartTrafic edge brain — same decision rules as platform/src/lib/algorithm.ts"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

Mode = Literal["normal", "school", "market", "night", "eco", "emergency", "failsafe"]
Color = Literal["red", "amber", "green", "flashing_amber", "off"]
Axis = Literal["ns", "ew"]


@dataclass
class Counts:
    m100: int = 0
    m200: int = 0
    m300: int = 0
    wait_s: float = 0.0
    motos: int = 0
    cars: int = 0
    buses: int = 0
    trucks: int = 0
    peds: int = 0
    emergency: bool = False


@dataclass
class Timing:
    min_green_s: int = 8
    max_green_s: int = 60
    yellow_s: int = 3
    all_red_s: int = 1
    min_ped_s: int = 12
    extension_s: int = 3


@dataclass
class ApproachState:
    name: str
    heading_deg: int = 0
    counts: Counts = field(default_factory=Counts)
    color: Color = "red"
    green_elapsed_s: float = 0.0
    ped_waiting: bool = False


def approach_score(c: Counts, mode: Mode) -> float:
    score = (
        c.m100 * 1.0
        + c.m200 * 1.6
        + c.m300 * 2.4
        + c.wait_s * 0.08
        + c.motos * 0.45
        + c.buses * 2.2
        + c.trucks * 2.8
        + c.peds * 1.8
    )
    if mode == "school":
        score += c.peds * 2.5
    if mode == "night" and (c.m100 + c.m200 + c.m300) == 0:
        score *= 0.2
    if c.emergency:
        score += 10_000
    return score


def axis_of(heading: int) -> Axis:
    d = heading % 360
    return "ew" if d in (90, 270) else "ns"


def _failsafe(approaches: list[ApproachState]) -> list[ApproachState]:
    for a in approaches:
        a.color = "flashing_amber"
    return approaches


def _set_axis(approaches: list[ApproachState], axis: Axis) -> list[ApproachState]:
    for a in approaches:
        a.color = "green" if axis_of(a.heading_deg) == axis else "red"
        if axis_of(a.heading_deg) != axis:
            a.green_elapsed_s = 0
    return _assert(approaches)


def _assert(approaches: list[ApproachState]) -> list[ApproachState]:
    greens = [a for a in approaches if a.color == "green"]
    axes = {axis_of(g.heading_deg) for g in greens}
    if len(axes) > 1:
        return _failsafe(approaches)
    return approaches


def next_phase(approaches: list[ApproachState], mode: Mode, timing: Timing) -> list[ApproachState]:
    if mode == "failsafe":
        return _failsafe(approaches)

    emergency = next((a for a in approaches if a.counts.emergency), None)
    if emergency:
        return _set_axis(approaches, axis_of(emergency.heading_deg))

    greens = [a for a in approaches if a.color == "green"]
    if not greens:
        winner = max(approaches, key=lambda a: approach_score(a.counts, mode))
        for a in approaches:
            a.green_elapsed_s = 0
        return _set_axis(approaches, axis_of(winner.heading_deg))

    current = axis_of(greens[0].heading_deg)
    group = [a for a in approaches if axis_of(a.heading_deg) == current]
    other = [a for a in approaches if axis_of(a.heading_deg) != current]
    elapsed = max(a.green_elapsed_s for a in group) + 1
    empty = all((a.counts.m100 + a.counts.m200 + a.counts.m300) == 0 for a in group)
    other_score = sum(approach_score(a.counts, mode) for a in other)
    group_score = sum(approach_score(a.counts, mode) for a in group)
    ped_hold = any(a.ped_waiting for a in group) and elapsed < timing.min_ped_s
    incoming = any(a.counts.m200 or a.counts.m300 for a in group)

    if ped_hold:
        for a in group:
            a.green_elapsed_s = elapsed
        return _assert(approaches)

    if (elapsed >= timing.min_green_s and empty and other_score > 2) or (
        elapsed >= timing.max_green_s and other_score >= group_score
    ):
        for a in approaches:
            a.green_elapsed_s = 0
        return _set_axis(approaches, "ew" if current == "ns" else "ns")

    for a in group:
        a.green_elapsed_s = elapsed + (timing.extension_s if incoming and not empty else 0)
    return _assert(approaches)
